Source/modules/webcl/WebCLProgram.cpp

```diff
@@ -0,0 +1,502 @@
+            if (!webCLBuildOptionsVector[i].startsWith("-")) {
+                es.throwWebCLException(WebCLException::INVALID_BUILD_OPTIONS, WebCLException::invalidBuildOptionsMessage);
+                return;
+            }
+
+            if (webCLSupportedBuildOptions.contains(AtomicString(webCLBuildOptionsVector[i])))
+                continue;
+
+            if (webCLBuildOptionsVector[i].startsWith(buildOptionDashD)) {
+                size_t j;
+                for (j = i + 1; j < webCLBuildOptionsVector.size() && !webCLBuildOptionsVector[j].startsWith("-"); ++j) {}
+                if (webCLBuildOptionsVector[i].stripWhiteSpace() == buildOptionDashD && j == i + 1) {
+                    es.throwWebCLException(WebCLException::INVALID_BUILD_OPTIONS, WebCLException::invalidBuildOptionsMessage);
+                    return;
+                }
+
+                i = --j;
+                continue;
+            }
+
+            es.throwWebCLException(WebCLException::INVALID_BUILD_OPTIONS, WebCLException::invalidBuildOptionsMessage);
+            return;
+        }
+    }
+
+    pfnNotify callbackProxyPtr = nullptr;
+    WebCLProgram* program = nullptr;
+    if (callback) {
+        if (m_callback) {
+            es.throwWebCLException(WebCLException::INVALID_OPERATION, WebCLException::invalidOperationMessage);
+            return;
+        }
+
+        // Store the callback, eventList to HashTable and call callbackProxy.
+        m_callback = callback;
+        callbackProxyPtr = &callbackProxy;
+        program = this;
+    }
+
+    cl_int err = CL_SUCCESS;
+    Vector<cl_device_id> clDevices;
+    Vector<RefPtr<WebCLDevice>> contextDevices = context()->getDevices();
+    if (devices.size()) {
+        Vector<cl_device_id> inputDevices;
+        for (auto device : devices)
+            inputDevices.append(device->getDeviceId());
+
+        size_t contextDevicesLength = contextDevices.size();
+        for (size_t z, i = 0; i < inputDevices.size(); i++) {
+            // Check if the inputDevices[i] is part of programs WebCLContext.
+            for (z = 0; z < contextDevicesLength; z++) {
+                if (contextDevices[z]->getDeviceId() == inputDevices[i]) {
+                    break;
+                }
+            }
+
+            if (z == contextDevicesLength) {
+                es.throwWebCLException(WebCLException::INVALID_DEVICE, WebCLException::invalidDeviceMessage);
+                return;
+            }
+
+            clDevices.append(inputDevices[i]);
+        }
+    } else {
+        for (auto contextDevice : contextDevices)
+            clDevices.append(contextDevice->getDeviceId());
+    }
+
+    if (!clDevices.size()) {
+        es.throwWebCLException(WebCLException::INVALID_VALUE, WebCLException::invalidValueMessage);
+        return;
+    }
+
+    m_isProgramBuilt = true;
+    err = clBuildProgram(m_clProgram, clDevices.size(), clDevices.data(), buildOptions.utf8().data(), callbackProxyPtr, program);
```

Comment:
this code don't make sure |this| lifetime. If you wanna take advantage of ThreadSafeRefCounted, you should do as follows:
```
struct WebCLProgramHolder {
RefPtr<WebCLProgram> program;
}

WebCLProgramHolder* holder = new WebCLProgramHolder;
holder->program = this;
clBuildProgram(..., holder);

// don't forget to delete holder in the callback.
```